From fb42d248dd697cc063f3565df749e80aeae3ff38 Mon Sep 17 00:00:00 2001
From: Anssi Hannula <anssi@xbmc.org>
Date: Sat, 17 Mar 2012 03:35:38 +0200
Subject: [PATCH] hack: workaround for old incompatible PVR addon databases

This needs to be cleaned up for inclusion to upstream XBMC git.
Specifically, it needs to be more generic (i.e. not limited to the PVR
incompatibility nor to the addon db and possibly limited to upgrades
only (no drop()ing) or alternatively some backup should be made of the
old db.

Bugfixes to sqlitedataset.cpp should be separated and drop() needs to
be called in generic UpdateVersion() on failure to prevent continuing
failure.
---
 xbmc/addons/AddonDatabase.cpp     |   26 +++++++++++++++++++++++++-
 xbmc/dbwrappers/sqlitedataset.cpp |    5 ++++-
 2 files changed, 29 insertions(+), 2 deletions(-)

diff --git a/xbmc/addons/AddonDatabase.cpp b/xbmc/addons/AddonDatabase.cpp
index fe4e7b4..b0df1c8 100644
--- a/xbmc/addons/AddonDatabase.cpp
+++ b/xbmc/addons/AddonDatabase.cpp
@@ -40,7 +40,24 @@ CAddonDatabase::~CAddonDatabase()
 
 bool CAddonDatabase::Open()
 {
-  return CDatabase::Open();
+  bool ok = CDatabase::Open();
+  if (ok)
+  {
+    try
+    {
+      // perform a validation to see if this db was broken during an upgrade from
+      // an old PVR addon database
+      m_pDS->exec("select addon,version,optional from dependencies limit 1");
+    }
+    catch (...)
+    {
+      // broken db, drop and recreate - hack
+      try { m_pDB->drop(); } catch(...) { }
+      Close();
+      ok = CDatabase::Open();
+    }
+  }
+  return ok;
 }
 
 bool CAddonDatabase::CreateTables()
@@ -119,11 +136,18 @@ bool CAddonDatabase::UpdateOldVersion(int version)
       m_pDS->exec("CREATE TABLE blacklist (id integer primary key, addonID text, version text)\n");
       m_pDS->exec("CREATE UNIQUE INDEX idxBlack ON blacklist(addonID)");
     }
+
+    // hack for old PVR databases - this will fail and prevent an incorrect update
+    // TODO: proper validation when upgrading databases
+    m_pDS->exec("select addon,version,optional from dependencies limit 1");
   }
   catch (...)
   {
     CLog::Log(LOGERROR, "Error attempting to upgrade an old addon database!");
     RollbackTransaction();
+    // hack - drop db or the unupgraded db will be left around and next upgrade will
+    // fail as well:
+    try { m_pDB->drop(); } catch (...) { }
     return false;
   }
   CommitTransaction();
diff --git a/xbmc/dbwrappers/sqlitedataset.cpp b/xbmc/dbwrappers/sqlitedataset.cpp
index 1833d6f..670167d 100644
--- a/xbmc/dbwrappers/sqlitedataset.cpp
+++ b/xbmc/dbwrappers/sqlitedataset.cpp
@@ -317,7 +317,10 @@ int SqliteDatabase::copy(const char *backup_name) {
 int SqliteDatabase::drop() {
   if (active == false) throw DbErrors("Can't drop database: no active connection...");
   disconnect();
-  if (!unlink(db.c_str())) {
+  CStdString db_fullpath;
+  URIUtils::AddFileToFolder(host, db, db_fullpath);
+
+  if (unlink(db_fullpath.c_str())) {
      throw DbErrors("Can't drop database: can't unlink the file %s,\nError: %s",db.c_str(),strerror(errno));
      }
   return DB_COMMAND_OK;
-- 
1.7.9.3

